client/src/components/scatterplot/scatterplot.js

```diff
@@ -320,48 +218,167 @@ class Scatterplot extends React.PureComponent {
-    svg.selectAll("*").remove();
-
-    // the axes are much cleaner and easier now. No need to rotate and orient
-    // the axis, just call axisBottom, axisLeft etc.
-    const xAxis = d3.axisBottom().ticks(7).scale(xScale);
-
-    const yAxis = d3.axisLeft().ticks(7).scale(yScale);
-
-    // adding axes is also simpler now, just translate x-axis to (0,height)
-    // and it's alread defined to be a bottom axis.
-    svg
-      .append("g")
-      .attr("transform", `translate(0,${height})`)
-      .attr("class", "x axis")
-      .call(xAxis);
-
-    // y-axis is translated to (0,0)
-    svg
-      .append("g")
-      .attr("transform", "translate(0,0)")
-      .attr("class", "y axis")
-      .call(yAxis);
-
-    // adding label. For x-axis, it's at (10, 10), and for y-axis at (width, height-10).
-    svg
-      .append("text")
-      .attr("x", 10)
-      .attr("y", 10)
-      .attr("class", "label")
-      .style("font-style", "italic")
-      .text(scatterplotYYaccessor);
-
-    svg
-      .append("text")
-      .attr("x", width)
-      .attr("y", height - 10)
-      .attr("text-anchor", "end")
-      .attr("class", "label")
-      .style("font-style", "italic")
-      .text(scatterplotXXaccessor);
+  fetchAsyncProps = async (props) => {
+    const {
+      scatterplotXXaccessor,
+      scatterplotYYaccessor,
+      colors: colorsProp,
+      crossfilter,
+      pointDilation,
+    } = props.watchProps;
+
+    const [
+      expressionXDf,
+      expressionYDf,
+      colorDf,
+      pointDilationDf,
+    ] = await this.fetchData(
+      scatterplotXXaccessor,
+      scatterplotYYaccessor,
+      colorsProp,
+      pointDilation
+    );
+    const colorTable = this.updateColorTable(colorsProp, colorDf);
+
+    const xCol = expressionXDf.icol(0);
+    const yCol = expressionYDf.icol(0);
+    const xScale = getXScale(xCol, 0, width);
+    const yScale = getYScale(yCol, height, 0);
+    const positions = this.computePointPositions(
+      xCol.asArray(),
+      yCol.asArray(),
+      xScale,
+      yScale
+    );
+
+    const colors = this.computePointColors(colorTable.rgb);
+
+    const { colorAccessor } = colorsProp;
+    const colorByData = colorDf?.col(colorAccessor)?.asArray();
+    const {
+      metadataField: pointDilationCategory,
+      categoryField: pointDilationLabel,
+    } = pointDilation;
+    const pointDilationData = pointDilationDf
+      ?.col(pointDilationCategory)
+      ?.asArray();
+    const flags = this.computePointFlags(
+      crossfilter,
+      colorByData,
+      pointDilationData,
+      pointDilationLabel
+    );
+
+    return {
+      positions,
+      colors,
+      flags,
+      width,
+      height,
+      xScale,
+      yScale,
+    };
+  };
+
+  createXQuery(geneName) {
+    const { annoMatrix } = this.props;
+    const { schema } = annoMatrix;
+    const varIndex = schema?.annotations?.var?.index;
+    if (!varIndex) return null;
+    return [
+      "X",
```

Comment:
nit: magic constant `"X"`

Is there a reason this is uppercase instead of lowercase?

I'm just thinking to avoid people accidentally use `"x"` instead of `"X"`, we could either use a constant `const X = "X"` or have a convention of always using lowercase `"x"` to prevent possible bug `"x" !== "X"`?

Reply:
this is not a magic or arbitrary variable name.  Please read: https://scanpy.readthedocs.io/en/stable/usage-principles.html or https://anndata.readthedocs.io/en/stable/

Reply:
resolved offline.